Add SPI and RTC support for 96Boards Wistrio

Mani-Sadhasivam opened the pull request:
This PR adds SPI1 and RTC support for 96Boards Wistrio board. It has been forked out of #18998 to ease merging the STM32 specifics independent of the core LoRa code.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
soc: arm: st_stm32: stm32l1: Add RTC/Counter support

Add RTC/Counter support for STM32L1 SoCs.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
boards: arm: 96b_wistrio: Enable on-chip RTC driver

Enable on-chip RTC driver for 96Boards Wistrio board.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
boards: arm: 96b_wistrio: Enable SPI1

Enable SPI1 available on the expansion header of 96Boards
Wistrio board.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

karlp commented on soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb:
```diff
@@ -9,7 +9,7 @@ if SOC_STM32L151XB
 config SOC
 	string
-	default "stm32l151xb"
+	default "stm32l151xba"
```
This can't be right, there are both xxxxB and xxxxBA parts in the wild.  See https://www.st.com/en/microcontrollers-microprocessors/stm32l151-152.html  They're both even still marked as ACTIVE.

See https://www.st.com/content/st_com/en/products/microcontrollers-microprocessors/stm32-32-bit-arm-cortex-mcus/stm32-ultra-low-power-mcus/stm32l1-series/stm32l151-152/stm32l151rb.html vs https://www.st.com/content/st_com/en/products/microcontrollers-microprocessors/stm32-32-bit-arm-cortex-mcus/stm32-ultra-low-power-mcus/stm32l1-series/stm32l151-152/stm32l151rb-a.html

Mani-Sadhasivam replied:
> This can't be right, there are both xxxxB and xxxxBA parts in the wild.

I didn't say that `stm32l151xb` is deprecated. I assumed that in zephyr mainline, there is only one L1X SoC present and it is based on the Wistrio board. So the exact part number of the SoC is found to be `stm32l151xba`, hence I just modified it.

@erwango What is the criteria for `default` SoC name?

Mani-Sadhasivam replied:
> this isn't the default soc, the file you're editing here is "Kconfig.defconfig.stm32l151xb" you're changing it to add an "a" suffix.

I think I misunderstood the actual difference between those 2 SoCs. Anyway, I'v asked the board vendor to clarify the SoC part number as there seems to be different versions of RAK811 modules. If they confirm it is `STM32L151XB-A` then I'll do the change accordingly.

Mani-Sadhasivam replied:
The `stm32flash` utility confirms that it is `STM32L151XB-A`.

erwango replied:
> @erwango What is the criteria for default SoC name?

So, first it should match your SoC. Then, it should match (in lower case), the definition in `stm32cube/stm32l1xx/soc/stm32l1xx.h`.
It will be used in `stm32cube/CMakeLists.txt`

Mani-Sadhasivam replied:
@erwango So for the `STM32L151xBA` SoC, I should create `soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xba`, right? It will be an exact duplicate of `stm32l151xb` though.

erwango replied:
Yes, this is needed, but if I'm correct, @karlp did it already

karlp replied:
I actually added the x8-a, because that's the part I wanted for something else.  I raised it on the slack channel, what the future plans were for collecting all the variants, because you're goign to explode in boilerplate. response there was "just add them as you need them, we'll sort it out later"

The _defconfig_ might be, but it will refer to a different dtsi file, as they have different flash/ram to the non-a parts.  (and less eeprom writes too, but that's getting off topic...)

Mani-Sadhasivam replied:
> Yes, this is needed, but if I'm correct, @karlp did it already

@erwango He added `x8-a` but for me, `xb-a` is needed. So I'll create a new one!

karlp commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -351,7 +351,7 @@ static int spi_stm32_configure(struct device *dev,
 #endif
 
-#ifndef CONFIG_SOC_SERIES_STM32F1X
+#if !defined(CONFIG_SOC_SERIES_STM32F1X) && !defined(CONFIG_SOC_SERIES_STM32L1X)
```
This is the wrong fix.  L1 has the FRF bit, you're just missing this fix in the HAL: https://github.com/zephyrproject-rtos/hal_stm32/pull/22 (enabled inside my other PR: https://github.com/zephyrproject-rtos/zephyr/pull/19729/commits/5e5afa0c5520dc80518ccaf2338271de76e92c6e)

Mani-Sadhasivam committed:
soc: arm: stm32: Add support for STM32L151XB-A SoC

STM32L151XB-A SoC is almost similar to the STM32L151XB SoC except that
it has more RAM (32KiB). Hence add devicetree and Kconfig support.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
boards: arm: 96b_wistrio: Fix the SoC used on this board

The RAK811 module used on this board incorporates STM32L151CB-A SoC,
which has more RAM (32 KiB) compared to its companion STM32L151CB.
Hence, fix the doc, dts and Kconfig to include correct part number.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam commented:
@erwango Rebased!

erwango reviewed (changes requested):
@Mani-Sadhasivam can you update the board doc with your change please ? Otherwise this is fine.

dbkinder reviewed (approved):
+1 for doc changes

erwango reviewed (changes requested):
@Mani-Sadhasivam, request to update board doc with additional supported feature (SPI, RTC) still stands.

Mani-Sadhasivam commented:
@erwango Sure, will update.

Mani-Sadhasivam commented:
@erwango Done!
